### ISSUE TITLE
Allow to dead code elimination

### ISSUE THREAD

**Issue by szagi3891** [facebook/flow#3492]
```
type someType = {
 a : number,
 b : string | null
}
...

const ff = (data: someType): string => {
  if (data.c === null) {
    return 'field b is smpty';
  }

  return 'some value';
}
```

At this moment for flowtype this code is correct. Only that this code contains a logical error.
Flowtype could easily detect this error.

```
data.c === null
```
Left side is undefined (data.c is always undefined), right side is null. This condition always returns false.
So enough detect that the types specified for the operator compare are disjoint.

Another example:

```
type someStatusType = 'loading' | 'error' | ok';

const getLabel = (status: someStatusType) => string {
  if (status === 'loading') {
    return 'Loading';
  }
  if (status === 'error') {
    return 'Error label';
  }
  if (status === 'complete') {            //dead code
    return 'Complete status';
  } 
  return 'ok';
};
```

This dead code can be detected by this comparison :
```
status === 'complete'
```
This types are disjoint ( 'loading' | 'error' | ok' ===  'complete' will always be false).

**Comment by nmn** [comment]
Flow is a type checker, and doesn't compile your code. The type definitions are removed by Babel.
So Flow won't be adding Dead-Code Elimination.

However, Flow could throw a warning when you do obviously false or obviously true comparisons. Essentially letting you do some sort of manual dead-code detection. In the second example, you probably meant to test 'ok' instead of 'complete'. Showing you a warning instead of silently removing the dead code is probably a better outcome.

**Comment by szagi3891** [comment]
Yes, indeed, I mean just about to flowtype reported warnings. Messages will allowing to manually correct code. These messages will be very helpful for refactoring.

**Comment by nmn** [comment]
Changed title to make that clearer. Fingers crossed.

**Comment by szagi3891** [comment]
Thank you :)

**Comment by sfisol** [comment]
This issue makes me frightened before every refactoring.

**Comment by yachaka** [comment]
This would be great.
